WordPress/WordPressTest/Services/MediaCoordinatorTests.swift

```diff
@@ -4,21 +4,30 @@ import Nimble
 @testable import WordPress
 
 class MediaCoordinatorTests: XCTestCase {
-    private var context: NSManagedObjectContext!
+    private var contextManager: ContextManagerMock!
+    private var context: NSManagedObjectContext! {
+        contextManager.mainContext
+    }
 
     private var coordinator: MediaCoordinator!
 
     override func setUp() {
         super.setUp()
-        context = TestContextManager().mainContext
+        contextManager = ContextManagerMock()
+        contextManager.setUpAsSharedInstance()
         coordinator = MediaCoordinator(MediaServiceFactoryMock())
     }
 
     override func tearDown() {
         super.tearDown()
+        // Make sure the DispatchQueue.async calls made in
+        // `MediaCoordinator.notifyObserversForMedia` are executed before tests
+        // finish. The async calls may cause crash if they are executed outside
+        // of this test case, since they use the Core Data stack created in
+        // `setUp` function above.
+        wait(for: 0.3)
```

Comment:
Arbitrary waits in unit tests are problematic because:

- There is no guarantee the timeout will always be enough (unless one makes it real long) 
- They can unnecessarily slow the tests down 

Have you tried something like adding an observer block that toggles a flag which we can use to ensure `MediaCoordinator` finished doing its things?

https://github.com/wordpress-mobile/WordPress-iOS/blob/735cf3fdd8a71086a19b516f02826f402ed3d961/WordPress/Classes/Services/MediaCoordinator.swift#L471-L493

```swift
func test...() {
  let post = ...
  ...

  var coordinatorDone = false
  // Not sure about the type signature for the observer closure, but let's pretend it `() -> Void`
  _ = coordinator.addObserver({ _ in coordinatorDone = true }, forMediaFor: post)

  let isPushingAllPendingMedia = coordinator...

  ...

  expect(coordinatorDone).toEventually(beTrue())
}
```

Or, if we need to account for the fact that there are more media, we could use `var coordinatorCalledTimes = 0` , `{ _ in coordinatorCalledTimes += 1 }` then `expect(coordinatorCalledTimes).toEventually(equal(/* expected times */))`

I haven't tried any of this code in the app, so apologies if it doesn't work out of the box.

Reply:
I thought about a similar approach, by using an `XCTExpectation`. But since this "waiting observers to be called" behaviour is needed for all tests, it's probably best to put the waiting in `tearDown`, rather than individual test. But fullling an expectation, or making an assertion like you suggested, in `tearDown` looks a bit weird to me.

Also, a tricky thing here is, we need to wait for _all observers_ to be notified, not just _a observer_ to be notified, so even though we can add our own observer and wait til it's called, we can't garentee _all other observers_ have been called.

Reply:
I've spent a bit more time looking into this, I think I might have been overthinking about this issue. It appears adding `setUpAsSharedInstance` and `tearDown` is sufficient, the additional waiting is not needed. For reference, this is the full change: https://github.com/wordpress-mobile/WordPress-iOS/commit/1690be7761a89d7eea1523ea485d31d99d058908

Reply:
BTW, I've added [a new commit](https://github.com/wordpress-mobile/WordPress-iOS/pull/18527/commits/cfa1aa697e15395b6792c31665ad2e6f0011dd6f) to remove the waiting code, I'd love to rebase and change the original commit, but I worry that may mess up the branch you created for https://github.com/wordpress-mobile/WordPress-iOS/pull/18528. I'm happy to rebase and clean up the commit history, if you are okay with changing the trial branch 😄

Reply:
Once a PR has been initially approved, we try to avoid rebasing, otherwise we need to re-review the entire PR (not just changes). 

This is helped by having small PRs 😅